src/ncp/ncp_base.cpp

```
@@ -825,6 +839,144 @@ otError NcpBase::SendQueuedResponses(void)
+        entry->mTid                 = SPINEL_HEADER_GET_TID(aHeader);
+        entry->mTransmitFrame.mPsdu = entry->mTransmitPsdu;
+        SuccessOrExit(error = DecodeStreamRawTxRequest(entry->mTransmitFrame));
+        break;
+
+    case kPendingCommandTypeEnergyScan:
+        entry->mScanChannel = aScanChannel;
+        break;
+
+    default:
+        ExitNow(error = OT_ERROR_NOT_FOUND);
+    }
+
+    mPendingCommandQueueTail++;
+
+exit:
+    return error;
+}
+
+otError NcpBase::HandlePendingTransmit(PendingCommandEntry *entry)
+{
+    otError       error = OT_ERROR_NONE;
+    otRadioFrame *frame;
+
+    mCurTransmitTID = entry->mTid;
+    mCurTransmitIID = entry->mIid;
+    VerifyOrExit(otLinkRawIsEnabled(mInstance), error = OT_ERROR_INVALID_STATE);
+    frame = otLinkRawGetTransmitBuffer(mInstance);
+    VerifyOrExit(frame != nullptr, error = OT_ERROR_NO_BUFS);
+    memcpy(frame, (void *)&entry->mTransmitFrame, sizeof(frame));
+    SuccessOrExit(error = otLinkRawTransmit(mInstance, &NcpBase::LinkRawTransmitDone));
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        LinkRawTransmitDone(&entry->mTransmitFrame, nullptr, error);
+    }
+    return error;
+}
+
+otError NcpBase::HandlePendingEnergyScan(PendingCommandEntry *entry)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(mCurScanChannel == kInvalidScanChannel, error = OT_ERROR_INVALID_STATE);
+    SuccessOrExit(error = otLinkRawEnergyScan(mInstance, entry->mScanChannel, mScanPeriod, LinkRawEnergyScanDone));
+
+    mCurScanChannel = static_cast<int8_t>(entry->mScanChannel);
+    mCurTransmitIID = entry->mIid;
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        if (mEncoder.BeginFrame(SPINEL_HEADER_FLAG | static_cast<uint8_t>(entry->mIid << SPINEL_HEADER_IID_SHIFT),
+                                SPINEL_CMD_PROP_VALUE_IS, SPINEL_PROP_MAC_SCAN_STATE) == OT_ERROR_NONE)
+        {
+            if (mEncoder.WriteUint8(SPINEL_SCAN_STATE_IDLE) == OT_ERROR_NONE)
```

If some error occurs in the middle of the construction of this frame, the encoder will be left at an intermediate state and the error will be ignored. Will that cause some problems?

---

I think that makes sense, the error is still ignored but it's OK as `LinkRawEnergyScanDone()` is already ignoring it.

BTW, active scans are async too. Do they need to be handled the same way?

---

For the RCP, active scans are handled up on the host using the transmit primitives.  I don't think we can handle it the same way on the RCP.